--- a/drivers/media/pci/intel/ipu-bridge.c	2024-03-15 21:19:29.000000000 +0300
+++ b/drivers/media/pci/intel/ipu-bridge.c	2024-03-23 09:49:04.219603603 +0300
@@ -60,6 +60,8 @@ static const struct ipu_sensor_config ip
 	IPU_SENSOR_CONFIG("OVTIDB10", 1, 560000000),
 	/* GalaxyCore GC0310 */
 	IPU_SENSOR_CONFIG("INT0310", 0),
+	/* Omnivision ov08x40 */
+	IPU_SENSOR_CONFIG("OVTI08F4", 1, 400000000),
 };
 
 static const struct ipu_property_names prop_names = {
