--- a/drivers/media/pci/intel/ipu6/ipu6.h	2024-03-23 09:44:09.682358949 +0300
+++ b/drivers/media/pci/intel/ipu6/ipu6.h	2024-03-24 15:38:11.026480690 +0300
@@ -24,10 +24,10 @@ struct ipu6_bus_device;
 #define IPU6_NAME			"intel-ipu6"
 #define IPU6_MEDIA_DEV_MODEL_NAME	"ipu6"
 
-#define IPU6SE_FIRMWARE_NAME		"intel/ipu6se_fw.bin"
-#define IPU6EP_FIRMWARE_NAME		"intel/ipu6ep_fw.bin"
-#define IPU6_FIRMWARE_NAME		"intel/ipu6_fw.bin"
-#define IPU6EPMTL_FIRMWARE_NAME		"intel/ipu6epmtl_fw.bin"
+#define IPU6SE_FIRMWARE_NAME		"intel/ipu/ipu6se_fw.bin"
+#define IPU6EP_FIRMWARE_NAME		"intel/ipu/ipu6ep_fw.bin"
+#define IPU6_FIRMWARE_NAME		"intel/ipu/ipu6_fw.bin"
+#define IPU6EPMTL_FIRMWARE_NAME		"intel/ipu/ipu6epmtl_fw.bin"
 
 enum ipu6_version {
 	IPU6_VER_INVALID = 0,
